pkg/clusterapi/upgrader.go

```diff
@@ -52,18 +52,22 @@ type CAPIChangeDiff struct {
 	InfrastructureProvider *types.ComponentChangeDiff
 }
 
-func (c *CAPIChangeDiff) toChangeDiff() *types.ChangeDiff {
-	r := make([]*types.ComponentChangeDiff, 0, 4+len(c.BootstrapProviders))
-	r = append(r, c.CertManager, c.Core, c.ControlPlane, c.InfrastructureProvider)
-	for _, bootstrapChangeDiff := range c.BootstrapProviders {
+func ToChangeDiff(capiChangeDiff *CAPIChangeDiff) *types.ChangeDiff {
+	if capiChangeDiff == nil {
+		logger.V(1).Info("Nothing to upgrade for CAPI")
+		return nil
+	}
+	r := make([]*types.ComponentChangeDiff, 0, 4+len(capiChangeDiff.BootstrapProviders))
+	r = append(r, capiChangeDiff.CertManager, capiChangeDiff.Core, capiChangeDiff.ControlPlane, capiChangeDiff.InfrastructureProvider)
+	for _, bootstrapChangeDiff := range capiChangeDiff.BootstrapProviders {
 		b := bootstrapChangeDiff
 		r = append(r, &b)
 	}
 
 	return types.NewChangeDiff(r...)
 }
 
-func (u *Upgrader) capiChangeDiff(currentSpec, newSpec *cluster.Spec, provider providers.Provider) *CAPIChangeDiff {
+func CapiChangeDiff(currentSpec, newSpec *cluster.Spec, provider providers.Provider) *CAPIChangeDiff {
```

Comment:
I would vote to keep this method private
And have another one that return the same type as the others